modules/storages/app/common/storages/peripherals/connection_validators/nextcloud_validator.rb

```diff
@@ -0,0 +1,74 @@
+# frozen_string_literal: true
+
+#-- copyright
+# OpenProject is an open source project management software.
+# Copyright (C) the OpenProject GmbH
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License version 3.
+#
+# OpenProject is a fork of ChiliProject, which is a fork of Redmine. The copyright follows:
+# Copyright (C) 2006-2013 Jean-Philippe Lang
+# Copyright (C) 2010-2013 the ChiliProject Team
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See COPYRIGHT and LICENSE files for more details.
+#++
+
+module Storages
+  module Peripherals
+    module ConnectionValidators
+      class NextcloudValidator
+        # Class Level interface will be moved to a superclass/mixin once we do the OneDrive port
+        class << self
+          def validation_groups
+            @validation_groups ||= {}
+          end
+
+          def register_group(group_name, klass, when: ->(storage, *) { storage.configured? })
+            validation_groups[group_name] = { klass:, when: }
+          end
+        end
+
+        register_group :base_configuration, Nextcloud::BaseConfigurationValidator
+        register_group :authentication, Nextcloud::AuthenticationValidator,
+                       when: ->(_, result) { result.group(:base_configuration).non_failure? }
+        register_group :ampf_configuration, Nextcloud::AmpfConfigurationValidator,
+                       when: ->(storage, result) {
+                         result.group(:base_configuration).non_failure? && storage.automatic_management_enabled?
+                       }
+
+        def initialize(storage:)
+          @storage = storage
+        end
+
+        def validate
+          validation_groups.each_with_object(ValidatorResult.new) do |(key, group_metadata), result|
+            if group_metadata[:when].call(@storage, result)
+              result.add_group_result(key, group_metadata[:klass].new(@storage).call)
+            end
+          end
+        rescue KeyError
```

Comment:
When would a `KeyError` occur in the given code?

Reply:
It could, if for some reason we try to access a result that isn't ready yet. It is more of a safeguard than anything. =)

Reply:
> if for some reason we try to access a result that isn't ready yet

Wouldn't that have a skipped/pending status? Also is there any parallel execution involved?

> It is more of a safeguard than anything.

Not sure if I am a fan of this argumentation. If you want a safeguard, you should rescue from `StandardError` ;-P

Reply:
Pardon my insistance on this topic, but isn't an exception for typos during coding _good_ behaviour? So that you notice your typo quickly?

Also I think I still do not understand where exactly I have to make a typo to get a key error.

Anyways, you got my feedback on this, I trust you to come to the right conclusion. I am leaving it up to you to keep this `rescue` in or not. I have not yet been convinced ^^